File: packages/jw-core-js/src/versification.ts
/**
 * Versification mapping — minimal port of the Python `jw_core.versification`
 * module (Fase 46). Maps a (book, chapter, verseStart, verseEnd) tuple
 * between nwt / masoretic / lxx / vulgate using the same JSON catalog the
 * Python implementation ships.
 *
 * Idempotent: `toCanonical({...from: X, to: X})` returns the input
 * unchanged with `isDiscrepant=false`.
 *
 * Lossless on round-trip for cataloged entries.
 */

import catalog from "./versification_map.json" with { type: "json" };

export type Tradition = "nwt" | "masoretic" | "lxx" | "vulgate";

export type VersificationIssue =
  | "superscription"
  | "chapter_split"
  | "verse_split"
  | "verse_merge"
  | "chapter_renumber"
  | "verse_shift";

export interface VerseCoord {
  chapter: number;
  verseStart: number;
  verseEnd?: number | null;
}

export interface VersificationEntry {
  book: string;
  book_num: number;
  issue: VersificationIssue;
  nwt: VerseCoord;
  masoretic?: VerseCoord;
  lxx?: VerseCoord;
  vulgate?: VerseCoord;
  source: string;
  explanation: Record<"en" | "es" | "pt", string>;
}

interface RawCoord {
  chapter: number;
  verse_start: number;
  verse_end?: number | null;
}

interface RawEntry {
  book: string;
  book_num: number;
  issue: VersificationIssue;
  nwt: RawCoord;
  masoretic?: RawCoord;
  lxx?: RawCoord;
  vulgate?: RawCoord;
  source: string;
  explanation: Record<"en" | "es" | "pt", string>;
}

interface RawCatalog {
  version: string;
  discrepancies: RawEntry[];
}

function toCoord(raw: RawCoord | undefined): VerseCoord | undefined {
  if (!raw) return undefined;
  return {
    chapter: raw.chapter,
    verseStart: raw.verse_start,
    verseEnd: raw.verse_end ?? null,
  };
}

function normalizeEntry(raw: RawEntry): VersificationEntry {
  const out: VersificationEntry = {
    book: raw.book,
    book_num: raw.book_num,
    issue: raw.issue,
    nwt: toCoord(raw.nwt) as VerseCoord,
    source: raw.source,
    explanation: raw.explanation,
  };
  const masoretic = toCoord(raw.masoretic);
  const lxx = toCoord(raw.lxx);
  const vulgate = toCoord(raw.vulgate);
  if (masoretic) out.masoretic = masoretic;
  if (lxx) out.lxx = lxx;
  if (vulgate) out.vulgate = vulgate;
  return out;
}

let cachedEntries: readonly VersificationEntry[] | null = null;

export function loadCatalog(): readonly VersificationEntry[] {
  if (cachedEntries == null) {
    const raw = catalog as RawCatalog;
    cachedEntries = raw.discrepancies.map(normalizeEntry);
  }
  return cachedEntries;
}

export const CATALOG_VERSION: string = (catalog as RawCatalog).version;

const TRADITIONS: readonly Tradition[] = ["nwt", "masoretic", "lxx", "vulgate"];

function coordFor(
  entry: VersificationEntry,
  tradition: Tradition,
): VerseCoord | undefined {
  return entry[tradition];
}

export interface MappingResult {
  refBook: string;
  refBookNum: number;
  coord: VerseCoord;
  fromTradition: Tradition;
  toTradition: Tradition;
  isDiscrepant: boolean;
  rationale: string | null;
}

export interface ToCanonicalArgs {
  book: string;
  bookNum: number;
  chapter: number;
  verseStart: number;
  verseEnd?: number | null;
  fromTradition?: Tradition;
  toTradition: Tradition;
}

/**
 * Map a coordinate between numbering traditions.
 *
 * Throws RangeError when either tradition is unknown.
 */
export function toCanonical(args: ToCanonicalArgs): MappingResult {
  const fromTradition = args.fromTradition ?? "nwt";
  if (!TRADITIONS.includes(fromTradition)) {
    throw new RangeError(`Unknown fromTradition: ${fromTradition}`);
  }
  if (!TRADITIONS.includes(args.toTradition)) {
    throw new RangeError(`Unknown toTradition: ${args.toTradition}`);
  }

  const coord: VerseCoord = {
    chapter: args.chapter,
    verseStart: args.verseStart,
    verseEnd: args.verseEnd ?? null,
  };

  if (fromTradition === args.toTradition) {
    return {
      refBook: args.book,
      refBookNum: args.bookNum,
      coord,
      fromTradition,
      toTradition: args.toTradition,
      isDiscrepant: false,
      rationale: null,
    };
  }

  const entries = loadCatalog();
  let matched: VersificationEntry | null = null;
  for (const entry of entries) {
    if (entry.book_num !== args.bookNum) continue;
    const src = coordFor(entry, fromTradition);
    const dst = coordFor(entry, args.toTradition);
    if (!src || !dst) continue;
    if (src.chapter !== coord.chapter) continue;
    if (src.verseStart !== coord.verseStart) continue;
    matched = entry;
    break;
  }

  if (!matched) {
    return {
      refBook: args.book,
      refBookNum: args.bookNum,
      coord,
      fromTradition,
      toTradition: args.toTradition,
      isDiscrepant: false,
      rationale: null,
    };
  }

  const src = coordFor(matched, fromTradition) as VerseCoord;
  const dst = coordFor(matched, args.toTradition) as VerseCoord;

  const chapterDelta = dst.chapter - src.chapter;
  const verseDelta = dst.verseStart - src.verseStart;

  const newChapter = Math.max(0, coord.chapter + chapterDelta);
  const newStart = Math.max(0, coord.verseStart + verseDelta);
  const newEnd =
    coord.verseEnd != null ? Math.max(0, coord.verseEnd + verseDelta) : null;

  return {
    refBook: matched.book,
    refBookNum: matched.book_num,
    coord: { chapter: newChapter, verseStart: newStart, verseEnd: newEnd },
    fromTradition,
    toTradition: args.toTradition,
    isDiscrepant: true,
    rationale: matched.explanation.en,
  };
}

export type Language = "en" | "es" | "pt";

/** Trilingual rationale for the discrepancy at this coordinate. */
export function explain(args: ToCanonicalArgs & { language?: Language }): string | null {
  const fromTradition = args.fromTradition ?? "nwt";
  if (fromTradition === args.toTradition) return null;
  const result = toCanonical(args);
  if (!result.isDiscrepant) return null;
  const language: Language = args.language ?? "en";
  for (const entry of loadCatalog()) {
    if (entry.book_num !== args.bookNum) continue;
    const src = coordFor(entry, fromTradition);
    if (!src) continue;
    if (src.chapter === args.chapter && src.verseStart === args.verseStart) {
      return entry.explanation[language] || entry.explanation.en;
    }
  }
  return null;
}

File: packages/jw-cli/src/jw_cli/commands/versification.py
"""``jw versification`` — map a reference between numbering traditions.

Three subcommands:
  map      <ref> --from nwt --to masoretic     show the mapped coordinate
  explain  <ref> --from nwt --to masoretic     human-readable explanation
  list     [--book Joel]                       list catalog discrepancies
"""

from __future__ import annotations

import typer
from jw_core.parsers.reference import parse_reference
from jw_core.versification import explain as _explain
from jw_core.versification import load_catalog, to_canonical
from jw_core.versification.models import Tradition

versification_app = typer.Typer(
    help="Map Bible references between numbering traditions (Fase 46).",
    no_args_is_help=True,
)


def _parse_or_die(ref: str) -> tuple[str, int, int, int, int | None]:
    parsed = parse_reference(ref)
    if parsed is None:
        raise typer.BadParameter(f"could not parse reference: {ref!r}")
    return (
        parsed.book_canonical,
        parsed.book_num,
        parsed.chapter,
        parsed.verse_start or 1,
        parsed.verse_end,
    )


@versification_app.command("map")
def map_cmd(
    ref: str = typer.Argument(..., help="Bible reference, e.g. 'Joel 2:28'."),
    from_tradition: Tradition = typer.Option(
        "nwt", "--from", help="Source tradition."
    ),
    to_tradition: Tradition = typer.Option(
        ..., "--to", help="Target tradition."
    ),
) -> None:
    """Print the mapped coordinate plus a one-line rationale."""

    book, book_num, chapter, verse_start, verse_end = _parse_or_die(ref)
    result = to_canonical(
        book=book,
        book_num=book_num,
        chapter=chapter,
        verse_start=verse_start,
        verse_end=verse_end,
        from_tradition=from_tradition,
        to_tradition=to_tradition,
    )
    end = (
        f"-{result.coord.verse_end}"
        if result.coord.verse_end is not None
        and result.coord.verse_end != result.coord.verse_start
        else ""
    )
    typer.echo(
        f"{result.ref_book} {result.coord.chapter}:"
        f"{result.coord.verse_start}{end} ({to_tradition})"
    )
    if result.rationale:
        typer.echo(result.rationale)


@versification_app.command("explain")
def explain_cmd(
    ref: str = typer.Argument(..., help="Bible reference, e.g. 'Psalm 51:1'."),
    from_tradition: Tradition = typer.Option("nwt", "--from"),
    to_tradition: Tradition = typer.Option(..., "--to"),
    language: str = typer.Option("en", "--lang", help="en|es|pt"),
) -> None:
    """Print the trilingual explanation for the discrepancy (if any)."""

    book, book_num, chapter, verse_start, verse_end = _parse_or_die(ref)
    out = _explain(
        book=book,
        book_num=book_num,
        chapter=chapter,
        verse_start=verse_start,
        verse_end=verse_end,
        from_tradition=from_tradition,
        to_tradition=to_tradition,
        language=language,  # type: ignore[arg-type]
    )
    typer.echo(out if out else "(no discrepancy — identical reference)")


@versification_app.command("list")
def list_cmd(
    book: str | None = typer.Option(
        None, "--book", help="Filter by canonical book name."
    ),
) -> None:
    """List catalog entries, optionally filtered by book."""

    catalog = load_catalog()
    if book:
        catalog = [e for e in catalog if e.book.lower() == book.lower()]
    typer.echo(f"{len(catalog)} entries")
    for e in catalog:
        nwt = e.nwt
        typer.echo(
            f"  {e.book} {nwt.chapter}:{nwt.verse_start}"
            + (f"-{nwt.verse_end}" if nwt.verse_end else "")
            + f"  [{e.issue}]"
        )
